### PR TITLE
Migrate to ESLint

### LINKED ISSUE #1295: Migrate to Eslint
tslint is deprecated.

### PR DESCRIPTION
**Pull request by leonardssh**
## Description

Migrate to ESLint

## Related Issue

https://github.com/inversify/InversifyJS/issues/1295

## How Has This Been Tested?

Using the following command:
```sh
npm run test
```

## Types of changes

- [x] Updated docs / Refactor code / Added a tests case (non-breaking change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I have updated the changelog.

### CODE REVIEW

**Review by PodaruDragos — CHANGES_REQUESTED** [review]
this is a bit too much. 
first the indententation seems to be changed.
second you also added preetier ? 
you also changed the default for `"`

i suggest leaving the defaults for those kind of things. @dcavanagh  what do you think ?

**Comment by dcavanagh** [comment]
I agree with you @PodaruDragos I think we can do it without prettier. My company uses airbnb base and modifies it slightly. That is what I put in there but I am open to changes. 

I personally like 2 spaces for tabs and single quote but that isn't necessary

**Comment by leonardssh** [comment]
Ok, then I can revert the last commit, and integrate ESLint without prettier. But, in my opinion, a little prettier config, gives a different look to the code, it make it look more cleaner. 😊

**Comment by PodaruDragos** [comment]
I personally like 2 spaces for tabs and single quote but that isn't necessary.

me too. I was just thinking that we won't alter the codebase. if you want do it. I am all in for that change 👍 
idk why but having 2 spaces feels more clean over 4. might be just me. I'll review this again if we change to 2 spaces.

i'm ok with w/e you decide here @dcavanagh @LeonardSSH  :)

**Comment by PodaruDragos** [comment]
@dcavanagh @LeonardSSH 
is there any way we can set up a discord server for inversify ? so we can talk about this kind of things to not spam github issues ?
gitter seems to be down.

**Comment by leonardssh** [comment]
@PodaruDragos I started last night to create a discord server for Inversify. At the moment it is a bit empty, but I plan to configure it A.S.A.P. 

https://discord.gg/CZ22rEC6vK

File: package.json [hunk 1]
```diff
@@ -73,6 +73,7 @@
     "nyc": "15.1.0",
     "performance-now": "2.1.0",
     "publish-please": "5.5.2",
+    "queue-microtask": "^1.2.2",
```
Comment (PodaruDragos) [inline package.json]: what's up with this ?

Comment (leonardssh) [reply package.json]: It's required by `@typescript-eslint/parser`, but I can fix this.

image

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,22 @@
+root = true
+
+[*]
+charset = utf-8
+end_of_line = lf
+insert_final_newline = true
+trim_trailing_whitespace = true
+
+[*.{js,ts}]
+indent_size = 4
```
Comment (PodaruDragos) [inline .editorconfig]: did we not agree to change this to 2 ?

File: .prettierrc [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{
+	"$schema": "http://json.schemastore.org/prettierrc",
+	"endOfLine": "lf",
+	"quoteProps": "as-needed",
+	"semi": true,
+	"singleQuote": true,
+	"tabWidth": 4,
```
Comment (PodaruDragos) [inline .prettierrc]: did we not agree to change this to 2 ?

File: .editorconfig (hunk 2 above)
Comment (leonardssh) [reply .editorconfig]: Sure, why not.

File: .prettierrc (hunk 3 above)
Comment (leonardssh) [reply .prettierrc]: Yeap.

File: .eslintrc.json [hunk 4]
```diff
@@ -0,0 +1,152 @@
+{
```
Comment (dcavanagh) [inline .eslintrc.json]: Will you move this back to .eslintrc.js.  Just so we can see the changed rules?

Thank you

**Comment by benedyktdryl** [comment]
@PodaruDragos would you mind reviewing changes again? It seems like merging this PR would be a good start - other PRs are kinda influenced by this one and I believe it would be great to merge this migration first.

**Comment by PodaruDragos** [comment]
> @PodaruDragos would you mind reviewing changes again? It seems like merging this PR would be a good start - other PRs are kinda influenced by this one and I believe it would be great to merge this migration first.

**Comment by PodaruDragos** [comment]
@benedyktdryl we have been talking a bit on discord about this and decided to just wait to merge the outstanding PR's then we can merge this to contain some of the merge conflicts.

**Comment by PodaruDragos** [comment]
@dcavanagh @LeonardSSH @benedyktdryl @notaphplover we will close this branch in favor of another were we takes things slower like: 

1. Update typescript with more restrictive options
2. Solve all the compiler errors
3. Create small commits which keeps the code with 0 compiler errors
4. Add eslint

**Pull request closed without merging**